onnxruntime/core/providers/webnn/builders/impl/attention_op_builder.cc

```diff
@@ -0,0 +1,568 @@
+                 \  Expand(G)           |    /                      |
+                  \   |     past_value  |   /                       |
+                qk_MatMul          \    |  /                        |
+                      |            ScatterND-----> present_value    |
+                  qk_Div              /                             |
+                      |              /                              |
+                     Add <----------/--------(attention_bias, one/finfo_min mask*)
+                      |            /
+                    Softmax     Expand(G)
+                       \         /
+                        \       /
+                      qkv_MatMul
+                             |
+                          Transpose (perm=0,2,1,3)
+                             |
+                          Reshape---(shape=B,S,W)
+                             |
+                           output
+*/
+
+Status GroupQueryAttentionOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                                           const logging::Logger& logger) const {
+  const auto& op_type = node.OpType();
+  if (op_type != "GroupQueryAttention") {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Unsupported attention op: ", op_type);
+  }
+
+  const auto& input_defs = node.InputDefs();
+
+  emscripten::val query_input = model_builder.GetOperand(input_defs[0]->Name());
+  emscripten::val key_input = model_builder.GetOperand(input_defs[1]->Name());
+  emscripten::val value_input = model_builder.GetOperand(input_defs[2]->Name());
+  emscripten::val past_key_input = model_builder.GetOperand(input_defs[3]->Name());
+  emscripten::val past_value_input = model_builder.GetOperand(input_defs[4]->Name());
+  emscripten::val seqlens_k_input = model_builder.GetOperand(input_defs[5]->Name());
+
+  std::vector<int64_t> input_q_shape, input_k_shape, input_v_shape, input_past_k_shape, input_past_v_shape;
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_q_shape, logger), "Cannot get query shape");
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[1], input_k_shape, logger), "Cannot get key shape");
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[2], input_v_shape, logger), "Cannot get value shape");
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[3], input_past_k_shape, logger), "Cannot get past_key shape");
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[4], input_past_v_shape, logger), "Cannot get past_value shape");
+  const auto q_rank = input_q_shape.size();
+  const auto k_rank = input_k_shape.size();
+  const auto v_rank = input_v_shape.size();
+  const auto past_k_rank = input_past_k_shape.size();
+  const auto past_v_rank = input_past_v_shape.size();
+  ORT_RETURN_IF_NOT(q_rank == 3 && k_rank == 3 && v_rank == 3,
+                    "The qkv shape should be batch_size * sequence_length * hidden_size.");
+  ORT_RETURN_IF_NOT(past_k_rank == 4 && past_v_rank == 4, "The qkv shape should be BNSH.");
```

Comment:
These checks should be move to `IsOpSupportedImpl` to earlier fallback the op if WebNN doesn't support it.